app/src/main/kotlin/com/saltedge/authenticator/features/connections/create/ConnectProviderInteractor.kt

```diff
@@ -0,0 +1,158 @@
+/*
+ * This file is part of the Salt Edge Authenticator distribution
+ * (https://github.com/saltedge/sca-authenticator-android).
+ * Copyright (c) 2021 Salt Edge Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 3 or later.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * For the additional permissions granted for Salt Edge Authenticator
+ * under Section 7 of the GNU General Public License see THIRD_PARTY_NOTICES.md
+ */
+package com.saltedge.authenticator.features.connections.create
+
+import com.saltedge.authenticator.app.ERROR_INVALID_AUTHENTICATION_DATA
+import com.saltedge.authenticator.app.ERROR_INVALID_DEEPLINK
+import com.saltedge.authenticator.app.ERROR_INVALID_RESPONSE
+import com.saltedge.authenticator.models.Connection
+import com.saltedge.authenticator.models.repository.ConnectionsRepositoryAbs
+import com.saltedge.authenticator.models.repository.PreferenceRepositoryAbs
+import com.saltedge.authenticator.sdk.api.model.GUID
+import com.saltedge.authenticator.sdk.api.model.connection.ConnectionStatus
+import com.saltedge.authenticator.sdk.tools.isReturnToUrl
+import com.saltedge.authenticator.sdk.tools.keystore.KeyStoreManagerAbs
+import com.saltedge.authenticator.sdk.tools.parseRedirect
+import com.saltedge.authenticator.sdk.v2.api.model.ConnectionID
+import com.saltedge.authenticator.sdk.v2.api.model.Token
+import com.saltedge.authenticator.sdk.v2.api.model.appLink.ConnectAppLinkDataV2
+import com.saltedge.authenticator.sdk.v2.api.model.error.ApiErrorData
+
+abstract class ConnectProviderInteractor(
+    private val keyStoreManager: KeyStoreManagerAbs,
+    private val preferenceRepository: PreferenceRepositoryAbs,
+    private val connectionsRepository: ConnectionsRepositoryAbs,
+) : ConnectProviderInteractorAbs {
+    override var contract: ConnectProviderInteractorCallback? = null
+    override var authenticationUrl: String = ""
+    override val hasConfigUrl: Boolean
+        get() = initialConnectData?.configurationUrl != null
+    override val hasConnection: Boolean
+        get() = connection.guid.isNotEmpty()
+    override val connectionName: String
+        get() = connection.name
+    override val geolocationRequired: Boolean?
+        get() = connection.geolocationRequired
+
+    private var initialConnectData: ConnectAppLinkDataV2? = null
```

Comment:
What should I do if the model changes and is very different from the model in v1?

Reply:
Now both are equal. Let's solve the issue when it will appear.